### PR TITLE
document context form/toolbar changes for 5.3

### PR DESCRIPTION
**Pull request by tylerkelly13**
Related Ticket: DOC-528

Description of Changes:
* Document changes to conflicting context toolbars and context forms
* Fix some other issues in that section
* Conditionalise the legacy content so it wont appear after the next major version release.

DOD:
- [x] nav.yml has been updated if applicable
- [x] file has been included where required if applicable
- [x] files removed have been deleted, not just excluded from the build if applicable

Review
- [x] Documentation Team Lead has reviewed
- [x] Product Manager has reviewed

### CODE REVIEW

File: _includes/context/positioning.md [hunk 1]
```diff
@@ -1,8 +1,17 @@
...
+### Positioning of context {{contextItemType}}
 
-* A `selection` position will place the ContextForm above or below the current selection, centred if possible.
+There are three options for positioning context {{contextItemType}}s: `selection`, `node`, or `line`.
 
-* A `line` position will place the ContextForm to the right (or left in Right-to-Left languages) of the current selection.
+* A `selection` position will place the context {{contextItemType}} above or below the current selection, centered if possible.
 
-* A `node` position will places the ContextForm
+* A `node` position will place the context {{contextItemType}} relative to the bounds of a node (e.g. a table or image). It applies to a selected node that does not match the selection due to CSS properties (like float).
```
Comment (metricjs) [inline _includes/context/positioning.md:15]: I'm confused by the second sentence?

File: _includes/context/positioning.md [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
+
+* A `line` position will place the context {{contextItemType}} to the right (or left in Right-to-Left languages) of the current selection.
```
Comment (metricjs) [inline _includes/context/positioning.md:17]: This feels... a little inaccurate 🤔 I should check it

File: _includes/context/priority.md [hunk 3]
```diff
@@ -1,3 +1,97 @@
...
+### Troubleshooting context toolbar and context form conflicts
+
+There are situations where custom context toolbars or custom context forms may conflict with:
+
+  - Context toolbars or context forms provided by the editor,
+  - Context toolbars or context forms provided by the [Quick Toolbars (`quickbars`) plugin]({{site.baseurl}}/plugins/quickbars/),
+  - Other custom context toolbars or custom context forms.
+
+{% if site.productmajorversion < 6 %}
+How these conflicts are resolves depends on the {{site.productname}} version.
```
Comment (metricjs) [inline _includes/context/priority.md:12]: s/resolves/resolved

File: _includes/context/priority.md [hunk 4]
```diff
@@ -1,3 +1,97 @@
...
+
+- For {{site.productname}} 5.3 or newer, see: [Determining the display priority of context toolbars and context forms]().
+- For {{site.productname}} 5.0, 5.1, or 5.2; see: [Determining the display priority of context toolbars and context forms (Legacy behavior)]().
+{% endif %}
+
+#### Determining the display priority of context toolbars and context forms
+
+There are three settings that determine the priority of context toolbars and context forms: `scope`, `predicate`, and `position`.
+
+- `scope` - Sets the context toolbar or form as visible of certain content (`node`) or a general (or global) toolbar or form (`editor`).
```
Comment (metricjs) [inline _includes/context/priority.md:22]: "as visible of certain content..."??

File: _includes/context/priority.md [hunk 5]
```diff
@@ -1,3 +1,97 @@
...
+- `predicate` - Sets a function that tests in the current context position against one or more conditions, returning a boolean value.
```
Comment (metricjs) [inline _includes/context/priority.md:23]: Is the "Sets" necessary? Consistency, yes, but it reads a bit weird to me

File: _includes/context/priority.md [hunk 6]
```diff
@@ -1,3 +1,97 @@
...
+- `position` - Sets where the context toolbar or form is rendered, relative to the current section (`selection`, `node`, or `line`).
+
+Generally:
+
+- Context _forms_ are prioritized over context _toolbars_.
+- Context _forms_ with `scope: 'node'` are prioritized over `scope: 'editor'`.
+- Only one context _form_ can be shown for a selection or cursor position, they cannot be concatenated to another context _form_ or a context _toolbar_.
+- Context _toolbars_ with `position: 'selection'` are prioritized over `position: 'node'`, and `position: 'line'` is given the lowest priority.
+- When the editor concatenate context _toolbars_ where there is more than one context _toolbar_ to display with:
```
Comment (metricjs) [inline _includes/context/priority.md:32]: You've got extra words in this or something

File: _includes/context/priority.md [hunk 7]
```diff
@@ -1,3 +1,97 @@
...
+  
+    - A matching predicate,
+    - The same `scope` and `position` values.
+
+  Concatenated toolbars may contain duplicate toolbar items.
+
+- If no matching context toolbars or context forms are found for the selection or cursor position, then editor will recursively search for matches on the parent node of the current node, until it reaches the root node of the editor content.
```
Comment (metricjs) [inline _includes/context/priority.md:39]: This is a good way to explain this!

File: _includes/context/priority.md [hunk 8]
```diff
@@ -1,3 +1,97 @@
...
+    If there any **context forms** found, the first one found will be displayed in the editor and the process will end.
+
+5. Find all **context toolbars** with both:
+
+    - Any `scope`
+    - A predicate matching the **parent node** of the current node, selection, or cursor position.
+
+    If there any matching **context toolbars**, the editor will prioritize the context toolbars based on the `position` value.
+
+    - All context toolbars with `position: 'selection'` or `position: 'node'` will be concatenated, the concatenated toolbar will be displayed, and the process will end.
```
Comment (metricjs) [inline _includes/context/priority.md:85]: This is the point at which it won't combine position selection and node toolbars - selection first, then if none found try node, then if none found try line. Different to step 3.

File: _includes/context/positioning.md (hunk 1 above)
Comment (tylerkelly13) [reply _includes/context/positioning.md:15]: yeah, old docs content I didn't review. Deleted the second sentence.

File: _includes/context/priority.md (hunk 3 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:12]: Fixed

File: _includes/context/priority.md (hunk 5 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:23]: yeah, removed.

File: _includes/context/priority.md (hunk 6 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:32]: not sure what happened there. Fixed?

File: _includes/context/priority.md (hunk 8 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:85]: Ah, copied but forgot to adapt.

File: _includes/context/positioning.md (hunk 2 above)
Comment (metricjs) [reply _includes/context/positioning.md:17]: This is fine

File: _includes/context/priority.md [hunk 9]
```diff
@@ -1,3 +1,98 @@
...
+The following description can be used for troubleshooting the behavior of context toolbars and context forms.
+
+The editor will determine which context toolbars or context form will be shown using following process:
+
+1. Find all **context forms** with both:
+
+    - `scope: 'node'`
+    - A predicate matching the current selection or cursor position.
+
+    If there any matching **context forms**, the first one found will be displayed in the editor and the process will end.
```
Comment (ltrouton) [inline _includes/context/priority.md:52]: Should this be 'If there are any...'? Same sentence structure is present multiple times below as well.

File: _includes/context/positioning.md [hunk 10]
```diff
@@ -1,8 +1,17 @@
+{% case page.title %}
+  {% when "Context toolbar" %}
+    {% assign contextItemType = "toolbar" %}
+  {% when "Context forms" %}
+    {% assign contextItemType = "form" %}
+{% endcase %}
+<a class="anchor" id="positioningcontexttoolbars"></a><a class="anchor" id="positioningcontextforms"></a>
 
-There are three options for positioning: `selection`, `line`, or `node`.
+### Positioning of context {{contextItemType}}
```
Comment (lnewson) [inline _includes/context/positioning.md:9]: This is incorrect, as it should be `Positioning of context forms`, whereas atm it's `Positioning of context form`. So it should have a trailing `s` for both toolbars and forms.

File: _includes/context/positioning.md [hunk 11]
```diff
@@ -1,8 +1,17 @@
...
 
-* A `selection` position will place the ContextForm above or below the current selection, centred if possible.
+There are three options for positioning context {{contextItemType}}s: `selection`, `node`, or `line`.
 
-* A `line` position will place the ContextForm to the right (or left in Right-to-Left languages) of the current selection.
+* A `selection` position will place the context {{contextItemType}} above or below the current selection, centered if possible.
 
-* A `node` position will places the ContextForm
+* A `node` position will place the context {{contextItemType}} relative to the bounds of a node (e.g. a table or image).
```
Comment (lnewson) [inline _includes/context/positioning.md:15]: Should this also say above/below similar to `selection`, as it won't position to the side?

File: _includes/context/priority.md [hunk 12]
```diff
@@ -1,3 +1,98 @@
...
+- For {{site.productname}} 5.3 or newer, see: [Determining the display priority of context toolbars and context forms]().
+- For {{site.productname}} 5.0, 5.1, or 5.2; see: [Determining the display priority of context toolbars and context forms (Legacy behavior)]().
+{% endif %}
+
+#### Determining the display priority of context toolbars and context forms
+
+There are three settings that determine the priority of context toolbars and context forms: `scope`, `predicate`, and `position`.
+
+- `scope` - Sets the context toolbar or form as either: specific to certain types of content (`node`), or a general (global) toolbar or form (`editor`).
+- `predicate` - A function that tests in the current context position against one or more conditions, returning a boolean value.
```
Comment (lnewson) [inline _includes/context/priority.md:23]: This sentence is weird and doesn't read very well. Not sure how to improve it though, but the whole `in the current context position against one or more conditions` is the problematic part I think.

File: _includes/context/priority.md [hunk 13]
```diff
@@ -1,3 +1,98 @@
...
+- `position` - Sets where the context toolbar or form is rendered, relative to the current section (`selection`, `node`, or `line`).
```
Comment (lnewson) [inline _includes/context/priority.md:24]: `section` probably isn't a very clear term here. I'd change this to `selection` or `context` instead.

File: ui-components/contexttoolbar.md [hunk 14]
```diff
@@ -19,29 +19,16 @@ A context toolbar is registered by calling the `addContextToolbar` API in the re
...
-
+{% include context/positioning.md %}
 
 ### Example configuration
 
 This example shows how the quickbars plugin adds the standard selection context toolbar and an example of a custom toolbar for image alignment. The context toolbar will show whenever any content is selected.
 
 {% include codepen.html id="context-toolbar" height="600" tab="js" %}
 
+### Launching a context toolbar programmatically
```
Comment (lnewson) [inline ui-components/contexttoolbar.md:30]: I thought we weren't going to be documenting this and it was meant to be an internal-only thing for now?

Comment (lnewson) [reply ui-components/contexttoolbar.md:30]: @metricjs you might know more on this? however, I remember being told it shouldn't be documented when I first started.

File: _includes/context/priority.md (hunk 9 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:52]: :facepalm: thanks

Comment (tylerkelly13) [reply _includes/context/priority.md]: Fixed by https://github.com/tinymce/tinymce-docs/pull/1517/commits/957488d197547b25cda061bfce54563b2a350c73

File: _includes/context/positioning.md (hunk 10 above)
Comment (tylerkelly13) [reply _includes/context/positioning.md]: Added the `s` in https://github.com/tinymce/tinymce-docs/pull/1517/commits/343c4fd8c3d4cf158d9dc1ac98df01601665e184

File: _includes/context/positioning.md (hunk 11 above)
Comment (tylerkelly13) [reply _includes/context/positioning.md]: updated in https://github.com/tinymce/tinymce-docs/pull/1517/commits/343c4fd8c3d4cf158d9dc1ac98df01601665e184

File: ui-components/contexttoolbar.md (hunk 14 above)
Comment (tylerkelly13) [reply ui-components/contexttoolbar.md:30]: We can remove it if needed. It has been in the docs for more than 2 years: https://github.com/tinymce/tinymce-docs/blame/88f95d130a77a2bfce3b128a266c21a22d2d8ff9/ui-components/contexttoolbar.md#L24

File: _includes/context/priority.md (hunk 12 above)
Comment (tylerkelly13) [reply _includes/context/priority.md:23]: agreed. It's fine if you already know what the predicate does, which isn't helpful

File: ui-components/contexttoolbar.md (hunk 14 above)
Comment (lnewson) [reply ui-components/contexttoolbar.md:30]: Ignore then I guess :shrug:

Comment (tylerkelly13) [reply ui-components/contexttoolbar.md:30]: after discussion, deciding to leave the content in.

**Pull request merged**